tools/cache_creator/cache_creator.cpp

```diff
@@ -23,12 +23,165 @@
+                             /* NumPerLine = */ 16, /* BytesPerLine = */ 6);
+    uuid = uuid.drop_front(groupSize);
+    if (!uuid.empty())
+      os << '-';
+  }
+  assert(isValidHexUUIDStr(res));
+  return res;
+}
+
+static uint8_t hexDigitToNum(char c) {
+  assert((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f'));
+  return (c >= '0' && c <= '9') ? c - '0' : 10 + (c - 'a');
+}
+
+// =====================================================================================================================
+// Converts a UUID string into its binary representation.
+//
+// @param hexStr : The input hex string in the UUID format
+// @param [out] outUuid : The output buffer where de-serialized UUID gets written to
+// @returns : `true` on success, `false` when the passed `hexStr` is not valid
+bool hexStringToUuid(llvm::StringRef hexStr, llvm::MutableArrayRef<uint8_t> outUuid) {
+  assert(outUuid.size() == VK_UUID_SIZE);
+  if (!isValidHexUUIDStr(hexStr))
+    return false;
+
+  // Go over all hex digits and write out a nibble at a time.
+  // Updates `outUuid` so that the current byte is always at the front.
+  for (auto &&posAndHexDigit : llvm::enumerate(llvm::make_filter_range(hexStr, [](char c) { return c != '-'; }))) {
+    const uint8_t num = hexDigitToNum(posAndHexDigit.value());
+    if (posAndHexDigit.index() % 2 == 0) {
+      outUuid.front() = num << 4;
+    } else {
+      outUuid.front() |= num;
+      outUuid = outUuid.drop_front();
+    }
+  }
+
+  return true;
+}
+
+// =====================================================================================================================
+// Extracts cache hash from the given elf file. Looks for the llpc_cache_hash note in note sections.
+//
+// @param elfBuffer : The input elf file
+// @returns : The hash found on success, or error on failure
+llvm::Expected<Util::MetroHash::Hash> getElfCacheHash(llvm::MemoryBufferRef elfBuffer) {
+  auto elfObjectOrErr = llvm::object::ELF64LEObjectFile::create(elfBuffer);
+  if (auto err = elfObjectOrErr.takeError())
+    return std::move(err);
+
+  using ElfFileTy = llvm::object::ELF64LEFile;
+  const ElfFileTy &elfFile = elfObjectOrErr->getELFFile();
+
+  auto sectionsOrErr = elfFile.sections();
+  if (auto err = sectionsOrErr.takeError())
+    return std::move(err);
+
+  for (const auto &sectionHeader : *sectionsOrErr) {
+    if (sectionHeader.sh_type != llvm::ELF::SHT_NOTE)
+      continue;
+
+    auto sectionNameOrErr = elfFile.getSectionName(sectionHeader);
+    if (auto err = sectionNameOrErr.takeError()) {
+      llvm::consumeError(std::move(err));
+      continue;
+    }
+    if (!sectionNameOrErr->startswith(".note"))
+      continue;
+
+    llvm::Error notesError(llvm::Error::success());
+    auto noteRange = elfFile.notes(sectionHeader, notesError);
+    if (notesError)
+      return std::move(notesError);
+
+    for (const ElfFileTy::Elf_Note &note : noteRange) {
+      if (!note.getName().startswith("llpc_cache_hash"))
+        continue;
```

Comment:
Are you thinking of storing it in an extended private header? We could check that all elf files have the same llpc version number and emit a warning when there's a mismatch.
Ideally, the llpc version should be a part of the elf-level fingerprint and validated before loading cache in the driver, but I can think it might be useful to store it in the cache header to make diagnosing cache-related issues easier.

What do you think?

Reply:
At this point in the code, do you have access to "LLPC_MAJOR_VERSION" defined [here](https://github.com/GPUOpen-Drivers/xgl/blob/19df23b69b3e7117d0fb738eb2badb91846220a7/CMakeLists.txt#L351)?  

I was thinking  the cache creator could immediately say that it cannot build a cache because the elf is not the same version of LLPC that the driver will use.

Reply:
OK, sounds good to me.